### ISSUE TITLE
Support custom callbacks

### ISSUE THREAD

**Issue by terrytangyuan** [sql-machine-learning/elasticdl#1106]
Currently ElasticDL supports custom processing logic for prediction outputs via `PredictionOutputsProcessor` provided in model definition file. However, we may want to make this more flexible so that users can implement custom callbacks during different phases of ElasticDL job. For example, `PredictionOutputsProcessor` can be seen as a callback that is executed after prediction outputs are made. Another use case is to run some custom logic before training starts https://github.com/wangkuiyi/elasticdl/issues/1039.

Some references:
* [Keras callbacks](https://www.tensorflow.org/beta/guide/keras/custom_callback)
* [SessionRunHook](https://www.tensorflow.org/api_docs/python/tf/train/SessionRunHook) that can be used in [EstimatorSpec](https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec)

**Comment by chunyang-wen** [comment]
Should we create a callback class for ElasticDL? In ALPS, we use `TrainerHook` which is similar to `tf.keras.CallBack`.  In ElasticDL's worker, we have:
+ `training_process`
+ `evaluation_process`
+ `prediction_process`

And there may be other places we should add support for callback such as:
+ `on_train_begin`
+ `on_train_end`
+ `on_batch_begin`
+ `on_batch_end`

The places where users can insert their logics can vary as our user number increases and we support more use cases.

So Let's start early.

**Comment by terrytangyuan** [comment]
Yes, I think this interface should be similar to `tf.keras.CallBack` so users can implement it in model definition files in a familiar keras style. 
Some use cases we observed that we want to support are (we can add more later on):
* Callbacks used in testing that checks retry times and model parameter updates: https://github.com/sql-machine-learning/elasticdl/pull/1229
* Callback similar to `PredictionOutputsProcessor` that is executed after prediction outputs are made. 
* Callback to write additional summaries to TensorBoard after we completed each evaluation job, e.g. currently we are writing TensorBoard summaries in `EvaluationService.complete_task()`.
* Callback to modulate learning rate as we are currently doing in `MasterServicer._modulate_lr_if_needed`.
* Callback for logic to be executed before training starts such as starting embedding service.
* Callback to perform early stopping when certain conditions are met.

cc: @chunyang-wen

**Comment by terrytangyuan** [comment]
Note that not all built-in Keras callbacks support distributed training, `validate_callback()` can be used to check this：https://github.com/tensorflow/tensorflow/blob/4607de425c48632454f2f3e6962577f6798df934/tensorflow/python/keras/distribute/distributed_training_utils.py#L240

**Comment by workingloong** [comment]
"Callback similar to PredictionOutputsProcessor that is executed after prediction outputs are made." 
Why did we not use  `tf.keras.callback` to implement function like PredictionOutputsProcessor?

There are also some modules similar to callback in ElasticDL.
LearningRateScheduler:
https://github.com/sql-machine-learning/elasticdl/blob/151184859989452e7105de100c5478e1770031c4/elasticdl/python/common/lr_scheduler.py#L4
similar to Keras callback [LearningRateScheduler](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/LearningRateScheduler)

Save params into checkpoint file in ElasticDL:
https://github.com/sql-machine-learning/elasticdl/blob/151184859989452e7105de100c5478e1770031c4/elasticdl/python/ps/servicer.py#L216
similar to Keras callback [ModelCheckpoint](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint)

In my opinion, we can design a common callback API to implement PredictionOutputsProcessor and LearningRateScheduler. However, the same method likes `on_train_batch_end` of different callbacks may be executed by different instances in ElasticDL. For example, LearningRateScheduler needs to execute `on_train_batch_end` on the PS instance and the callback for early stop may need to execute it on the Master instance.

**Comment by terrytangyuan** [comment]
Yes those two items are already mentioned in my previous comments. Earlier we didn’t have a lot of use cases so we didn’t spend much time on designing this.
